### ISSUE TITLE
Connecting to Oracle error with oracledb

### ISSUE THREAD

**Issue by bnewman-tech** [sfu-db/connector-x#550]
#### What language are you using?
Python

#### What version are you using?
0.3.2 on a Macbook M1 Max Machine

#### What database are you using?
Oracle

#### What dataframe are you using?
Pandas and/or Polars

#### Can you describe your bug?
I am trying to use [oracledb](https://oracle.github.io/python-oracledb/) thin mode driver with [connector-x](https://github.com/sfu-db/connector-x) without installing the Oracle Client. Is this possible and if so, what I am doing wrong?

##### Example query / code
```
import connectorx as cx

ENGINE_PATH = "oracle+oracledb://username:password@host:post/service_name"
query = "select * from table"  # query string

cx.read_sql(ENGINE_PATH, query)
```

#### What is the error?

[2023-11-01T02:17:30Z ERROR r2d2] DPI Error: DPI-1047: Cannot locate a 64-bit Oracle Client library:
"dlopen(libclntsh.dylib, 0x0001): tried: 'libclntsh.dylib' (no such file),
'/System/Volumes/Preboot/Cryptexes/OSlibclntsh.dylib' (no such file), '/usr/lib/libclntsh.dylib' 
'(no such file, not in dyldcache), 'libclntsh.dylib' (no such file), '/usr/lib/libclntsh.dylib' 
(no such file, not in dyld cache)". See https://oracle.github.io/odpi/doc/installation.html#macos for help

**Comment by johalnes** [comment]
I'm wondering about the same thing! Any progress?

**Comment by bnewman-tech** [comment]
Good morning, 

I wanted to talk a moment to follow up on this and see if it is possible to use the thin driver https://oracle.github.io/python-oracledb/?

**Comment by pangjunrong** [comment]
Hi @bnewman-tech, I imagine you would be able to use it. However, if the thin driver is found within the oracledb lib that you installed into your virtual environment via pip, it will not be located when running the connector-x. By default, an Oracle connection searches for the driver first in its own module directory (connector-x lib) and expands into the directories defined in your PATH environment variable. Given the error you have provided, it is likely that the latter has not been set to point to the driver's filepath within the oracledb lib.

If you want to investigate which paths are scanned when looking for the driver, you can set a system environment variable as `DPI_DEBUG_LEVEL=64`. This will enable the DPI Error to be more verbose. You can [refer to this link](https://odpi-c.readthedocs.io/en/latest/user_guide/installation.html#oracle-client-library-loading) to read up more about how the driver location is searched.

Once the driver files are in the right location, you should be good to go!

**Comment by Feromond** [comment]
I am having the same issue. When I am trying to use connector-x in a cloud environment such as Databricks on their clusters with custom runtimes, I do not have the ability to add the required drivers. The thin driver would work well but it would have to come included with connector-x then. Is there any other solution that could work for environments where its not possible to go and install / add the oracle client drivers.

**Comment by ChenHuang-IIASA** [comment]
Any progress here?

**Comment by pangjunrong** [comment]
@ChenHuang-IIASA Have you tried setting ORACLE_HOME environment variable to the full directory path of the Oracle Thin Client?

**Comment by ChenHuang-IIASA** [comment]
Yes. It works. But I don't know if connectorx will support ultilizing the oracledb package directly.
For example, I prefer use uvx --with connectorx --with oracledb ipython to launch a temp query instance quickly without the requirement to install oracle client.

**Comment by pangjunrong** [comment]
To my knowledge, that is not possible. The `python-oracledb` library has a "thin" mode that directly implements the Oracle database protocol in Cython, bypassing the need for a client to interact with Oracle DBs. This is not a file-based driver we can just use, so it will not work.

The `connector-x`'s implementation currently uses `r2d2-oracle`, which is built on top of `rust-oracle`. The [runtime requirements](https://github.com/kubo/rust-oracle?tab=readme-ov-file#run-time-requirements) state the need for an Oracle Client 11.2 or later, and that needs to be fulfilled either using Oracle Instant Client or Oracle Client.

The only way I imagine for such drivers/clients to be omitted completely is if `rust-oracle` is rewritten with a "thin" mode option to directly implement the relevant database protocols, just like `python-oracledb`. You might find better answers by [opening an issue](https://github.com/kubo/rust-oracle/issues) with them to determine if that's feasible. I'll be closing this since it's been answered.
